va/x11/va_x11.c

```diff
@@ -77,8 +77,34 @@ static VAStatus va_DisplayContextGetDriverNames(
 {
     VAStatus vaStatus = VA_STATUS_ERROR_UNKNOWN;
 
-    if (!getenv("LIBVA_DRI3_DISABLE"))
+    if (!getenv("LIBVA_DRI3_DISABLE")) {
+        unsigned old_num_drivers = *num_drivers;
+
         vaStatus = va_DRI3_GetDriverNames(pDisplayContext, drivers, num_drivers);
+        /* As of 8 July 2023, i965 and iHD drivers lack DRI3 support.
+         *
+         * Requests by the community were raised as early as 29 July 2017,
+         * with DRI3 support landing in libva on the 27 September 2022. At of
+         * time of writing it's unknown if/when that would materialise.
+         *
+         * To handle this on libva level, we are explicitly disabling DRI3
+         * support on said drivers - it scales better than having every user
+         * to set the environment override listed above.
+         */
+        if (vaStatus == VA_STATUS_SUCCESS) {
```

Comment:
I think we need to have a way for iHD/i965 to go via DRI3 for those users which don't use vaPutSurface. I suggest to allow `LIBVA_DRI3_DISABLE=0|1` values:
* If user specified `LIBVA_DRI3_DISABLE=1`, then DRI3 gets disabled (regardless of the driver)
* If user specified `LIBVA_DRI3_DISABLE=0`, then DRI3 is not disabled (regardless of the driver)
* If user did not specify `LIBVA_DRI3_DISABLE`, then by default DRI3 is disabled for particular drivers (iHD and i965)

Something like this:
```
static VAStatus va_DisplayContextGetDriverNames(
    VADisplayContextP pDisplayContext,
    char **drivers, unsigned *num_drivers
)
{
    VAStatus vaStatus = VA_STATUS_ERROR_UNKNOWN;

    vaStatus = va_DRI3_GetDriverNames(pDisplayContext, drivers, num_drivers);
    
    const char* dri3_disable = getenv("LIBVA_DRI3_DISABLE");
    if (dri3_disable && atoi(dri3_disable)) {
        vaStatus = VA_STATUS_ERROR_UNKNOWN;
    }
    else if (!dri3_disable) {
        for (unsigned i = 0; i < *num_drivers; i++) {
            if (drivers[i] && (!strcmp(drivers[i], "iHD") ||
                                !strcmp(drivers[i], "i965")))
                vaStatus = VA_STATUS_ERROR_UNKNOWN;
        }
    }
    if (vaStatus == VA_STATUS_ERROR_UNKNOWN) {
        for (unsigned i = 0; i < *num_drivers; i++)
            free(drivers[i]);
        *num_drivers = old_num_drivers;
    }
    if (vaStatus != VA_STATUS_SUCCESS)
        vaStatus = va_DRI2_GetDriverNames(pDisplayContext, drivers, num_drivers);
#ifdef HAVE_NVCTRL
    if (vaStatus != VA_STATUS_SUCCESS)
        vaStatus = va_NVCTRL_GetDriverNames(pDisplayContext, drivers, num_drivers);
#endif
#ifdef HAVE_FGLRX
    if (vaStatus != VA_STATUS_SUCCESS)
        vaStatus = va_FGLRX_GetDriverNames(pDisplayContext, drivers, num_drivers);
#endif
    return vaStatus;
}
```

Reply:
Seems reasonable - will respin in the upcoming days.

Reply:
The above example always calls va_DRI3_GetDriverNames and does not properly handle num_drivers. MR updated to handle that + the 3-state requested.